python/tk_multi_publish2/api/tree.py

```diff
@@ -0,0 +1,227 @@
+
+        if item == self.root_item:
+            raise sgtk.TankError("Removing the root item is not allowed.")
+
+        # all other items should have a parent
+        item.parent.remove_child(item)
+
+    def save_file(self, file_path):
+        """Save the tree to disk at the supplied path."""
+
+        with open(file_path, "w") as file_obj:
+            try:
+                self.save(file_obj)
+            except Exception, e:
+                logger.error(
+                    "Erorr saving the publish tree to disk: %s" % (e,)
+                )
+                raise
+
+    def save(self, file_obj):
+        """
+        Write a json-serialized representation of the publish tree to the
+        supplied file-like object.
+        """
+        try:
+            json.dump(self, file_obj, indent=2, cls=_PublishTreeEncoder)
+        except Exception, e:
+            logger.error(
+                "Error saving publish tree: %s\n%s" %
+                (e, traceback.format_exc())
+            )
+            raise
+
+    def to_dict(self):
+        """
+        Returns a dictionary representation of the publish tree. Typically used
+        during serialization of the publish tree.
+        """
+        return {"root_item": self.root_item.to_dict()}
+
+    @property
+    def items(self):
+        """A depth-first generator of all items to be published."""
+        for item in self._tree.root_item:
+            yield item
+
+    @property
+    def persistent_items(self):
+        """Returns a list of persistent items in the tree."""
+        return [i for i in self.root_item.children if i.persistent]
+
+    @property
+    def root_item(self):
+        """Returns the root item of this tree."""
+        return self._root_item
+
+    def _format_tree(self, parent_item, depth=0):
+        """
+        Depth first traversal and string formatting of the tree given a root
+        node.
+
+        :param parent_item: The item to begin with
+        :param depth: The current depth of the traversal (used for indentation)
+        """
+
+        tree_str = ""
+
+        for item in parent_item.children:
+
+            # format the item itself
+            tree_str += "%s[item] %s\n" % ((depth * 2) * " ", str(item))
+
+            # now do the item's tasks
+            for task in item.tasks:
+                tree_str += "%s[task] %s\n" % (
+                    ((depth * 2) + 2) * " ", str(task))
+
+            # process the children as well
+            tree_str += "%s" % (self._format_tree(item, depth=depth+1),)
+
+        return tree_str
+
+
+class _PublishTreeEncoder(json.JSONEncoder):
+    """
+    Implements the json encoder interface for custom publish tree serialization.
+    """
+    def default(self, publish_tree):
+        return publish_tree.to_dict()
+
+
+class _PublishTreeDecoder(json.JSONDecoder):
+    """
+    Implements the json decoder interface for custom publish tree
+    deserialization.
+    """
+    def decode(self, json_str):
```

Comment:
When decoding, you read that number and you pass it as a parameter to from_dict so that more recent versions of the code know which version of the schema the document is?

Reply:
`def from_dict(schema_version=LATEST)` which is defined to whatever is the latest version.